mp2-common/src/storage_key.rs

```diff
@@ -28,42 +40,77 @@ const INPUT_TUPLE_LEN: usize = 2 * INPUT_ELEMENT_LEN;
 /// The whole padded length for the inputs
 const INPUT_PADDED_LEN: usize = PAD_LEN(INPUT_TUPLE_LEN);
 
+/// Wires associated with the MPT key from the Keccak computation of location
+// It's only used for mapping slot.
 #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
-/// Wires associated with the MPT key from the keccak computation of location
 pub struct KeccakMPTWires {
-    /// Actual keccak wires created for the computation of the "location" for
-    /// the storage slot
-    pub keccak_location: ByteKeccakWires<INPUT_PADDED_LEN>,
-    /// Actual keccak wires created for the computation of the final MPT key
-    /// from the location. THIS is the one to use to look up a key in the
+    /// Actual Keccak wires created for the computation of the base for the storage slot
+    pub keccak_base: ByteKeccakWires<INPUT_PADDED_LEN>,
+    /// Actual Keccak wires created for the computation of the final MPT key
+    /// from the location. this is the one to use to look up a key in the
     /// associated MPT trie.
     pub keccak_mpt_key: KeccakWires<{ PAD_LEN(HASH_LEN) }>,
-    /// The MPT key derived in circuit from the storage slot, in NIBBLES
+    /// The MPT key derived in circuit from the storage slot in nibbles
     /// TODO: it represents the same information as "exp" but in nibbles.
     /// It doesn't need to be assigned, but is used in the higher level circuits
     pub mpt_key: MPTKeyWire,
 }
 
+// The Keccak MPT computation
+// It's only used for mapping slot.
 struct KeccakMPT;
 
 impl KeccakMPT {
+    /// Build the Keccak MPT with no offset (offset = 0).
     fn build<F: RichField + Extendable<D>, const D: usize>(
         b: &mut CircuitBuilder<F, D>,
         inputs: VectorWire<Target, INPUT_PADDED_LEN>,
     ) -> KeccakMPTWires {
-        let keccak_location = KeccakCircuit::<{ INPUT_PADDED_LEN }>::hash_to_bytes(b, &inputs);
-        // keccak(location) - take the output and copy it in a slice large
-        // enough for padding.
-        let mut padded_location = [b.zero(); PAD_LEN(HASH_LEN)];
-        padded_location[0..HASH_LEN].copy_from_slice(&keccak_location.output.arr);
+        let keccak_base = KeccakCircuit::<{ INPUT_PADDED_LEN }>::hash_to_bytes(b, &inputs);
+        let location = keccak_base.output.arr;
+
+        Self::build_location(b, keccak_base, location)
+    }
+
+    /// Build the Keccak MPT with a specified offset of Uint32.
+    fn build_with_offset<F: SerializableRichField<D> + Extendable<D>, const D: usize>(
+        b: &mut CircuitBuilder<F, D>,
+        inputs: VectorWire<Target, INPUT_PADDED_LEN>,
+        offset: Target,
+    ) -> KeccakMPTWires {
+        // location = keccak(inputs) + offset
+        let keccak_base = KeccakCircuit::<{ INPUT_PADDED_LEN }>::hash_to_bytes(b, &inputs);
+        let base = keccak_base.output.arr.pack(b, Endianness::Big);
```

Comment:
Just noticed that `hash_to_bytes` decompose the output of Keccak in little-endian, while here you are packing it again in big-endian. Wouldn't be better to use `hash_vector`, which returns the output as u32 limbs which can be directly be employed to initialize a `UInt256Target`?

Reply:
It really depends how EVM is interpreting the output of Keccak as a U256 to add offset. I think your current way of initializing the `UInt256Target` makes sense, but maybe a range-check on each byte of the `hash_to_bytes` output is needed, as `hash_to_bytes` should be safe to use without range-checking the output bytes only if those bytes are then fed as input to another Keccak, which is not the case here. Wdyt?